### PR TITLE
Multi threaded bruteforce

### PR DESCRIPTION
**Pull request by ArtemiyGranat**

### CODE REVIEW

**Review by hnarkaytis — APPROVED** [review]

**Comment by ArtemiyGranat** [comment]
Tested bruteforce of 5-character password with 64-character alphabet on Debian desktop (16 cores, changed `QUEUE_SIZE` to 16 for this test).
Singlethreaded:
```
artemii@debian ~/C/brute (multi-threaded-bruteforce)> time ./main -h abC8QA7f5FZtA -l 5 -a ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/
Password found: hjklm
________________________________________________________
Executed in   35.29 mins    fish           external
   usr time   35.29 mins  516.00 micros   35.29 mins
   sys time    0.00 mins  216.00 micros    0.00 mins
   ```
Multithreaded:
```
artemii@debian ~/C/brute (multi-threaded-bruteforce)> time ./main -h abC8QA7f5FZtA -l 5 -a ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/ -m
Password found: hjklm
________________________________________________________
Executed in  843.78 secs    fish           external
   usr time   73.32 mins  448.00 micros   73.32 mins
   sys time   27.67 mins  187.00 micros   27.67 mins
```
Acceleration is 35.29 mins / 14.063 mins ~ 2,5 times. I guess it could be better especially when I use a 16-core machine, will test it on MacOS after `crypt.h` porting (I also changed `MAX_PASSWORD_LENGTH` to 30 when I tried to test 30-char password and forgot to change it back, but it shouldn't affect the singlethreaded/multithreaded ratio).

**Comment by ArtemiyGranat** [comment]
I found a trouble with my semaphore implementation so when I run multithreaded mode on macOS it stucks in infinite loop. But once in a while it doesn't stuck and prints `Password found` with correct password or `Password not found`. Maybe there is a race condition or something in my semaphore impl. Multithreaded mode on Linux is working correctly.

**Comment by ArtemiyGranat** [comment]
> I found a trouble with my semaphore implementation so when I run multithreaded mode on macOS it stucks in infinite loop. But once in a while it doesn't stuck and prints `Password found` with correct password or `Password not found`. Maybe there is a race condition or something in my semaphore impl. Multithreaded mode on Linux is working correctly.

UPD: It's trouble with `crypt.h` port.

**Comment by ArtemiyGranat** [comment]
> I found a trouble with my semaphore implementation so when I run multithreaded mode on macOS it stucks in infinite loop. But once in a while it doesn't stuck and prints `Password found` with correct password or `Password not found`. Maybe there is a race condition or something in my semaphore impl. Multithreaded mode on Linux is working correctly.

Fixed it I guess.

File: queue.c [hunk 1]
```diff
@@ -7,101 +7,109 @@ queue_init (queue_t *queue)
...
+  if (pthread_mutex_destroy (&queue->head_mutex) != 0)
+    return (S_FAILURE);
 
-  return S_SUCCESS;
-}
\ No newline at end of file
+  if (pthread_mutex_destroy (&queue->tail_mutex) != 0)
+    return (S_FAILURE);
+
+
```
Comment (hnarkaytis) [inline queue.c:113]: extra blank line

File: Makefile [hunk 2]
```diff
@@ -1,13 +1,27 @@
 .PHONY: clean
-CFLAGS=-O2 -Wall -Wpedantic -Wextra -pthread
-LIBS=-lcrypt
+CFLAGS=-O2 -Wall -Wpedantic -Wextra -pthread -ggdb3
 OBJ=brute.o common.o main.o multi.o queue.o single.o
 TARGET=main
 
+ifeq ($(shell uname), Linux)
```
Comment (hnarkaytis) [inline Makefile:6]: I guess we could unconditionally link with our local version of crypt_r

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (ArtemiyGranat) [reply Makefile:6]: > I guess we could unconditionally link with our local version of crypt_r

Builded program on Podman machine, `make` output is
```
[core@localhost brute]$ make
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o brute.o brute.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o common.o common.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o main.o main.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o multi.o multi.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o queue.o queue.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o single.o single.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3 brute.o common.o main.o multi.o queue.o single.o -lcrypt -o main
```

Comment (ArtemiyGranat) [reply Makefile:6]: Debian is the same:
```
artemii@debian ~/C/brute (main)> make
cc -O2 -Wall -Wpedantic -Wextra -pthread   -c -o brute.o brute.c
cc -O2 -Wall -Wpedantic -Wextra -pthread   -c -o common.o common.c
cc -O2 -Wall -Wpedantic -Wextra -pthread   -c -o main.o main.c
cc -O2 -Wall -Wpedantic -Wextra -pthread   -c -o multi.o multi.c
cc -O2 -Wall -Wpedantic -Wextra -pthread   -c -o queue.o queue.c
cc -O2 -Wall -Wpedantic -Wextra -pthread   -c -o single.o single.c
cc -O2 -Wall -Wpedantic -Wextra -pthread brute.o common.o main.o multi.o queue.o single.o -lcrypt -o main
```

Comment (ArtemiyGranat) [reply Makefile:6]: Debian build has the same output:
```
artemii@debian ~/C/brute (multi-threaded-bruteforce)> make
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o brute.o brute.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o common.o common.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o main.o main.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o multi.o multi.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o queue.o queue.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3   -c -o single.o single.c
cc -O2 -Wall -Wpedantic -Wextra -pthread -ggdb3 brute.o common.o main.o multi.o queue.o single.o -lcrypt -o main
```

Comment (hnarkaytis) [reply Makefile:6]: I mean - no need to do branching based on OS and always link with local copy of crypt_r.
Use ./crypt/*.[ha] for both Linux and MacOs.
	LIBS+=crypt/libcrypt.a
	CFLAGS+=-I./crypt

File: single.c [hunk 3]
```diff
@@ -3,14 +3,16 @@
 #include "brute.h"
 #include <string.h>
 
+#define crypt_r __crypt_r
```
Comment (hnarkaytis) [inline single.c:6]: Better to use weak_alias(__crypt_r, crypt_r);

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

File: single.c (hunk 3 above)
Comment (ArtemiyGranat) [reply single.c:6]: > Better to use weak_alias(__crypt_r, crypt_r);

There is no `weak_alias` on MacOS, I tried to change it to its define on Linux (`extern __typeof(__crypt_r) crypt_r __attribute__((weak, alias("__crypt_r")));` and got `error: aliases are not supported on darwin`. Are there any alternatives?

**Comment by ArtemiyGranat** [comment]
Once in a while password not found on MacOS, I think there is some trouble with `crypt_r` on MacOS, because brute works correctly all the time on Debian and Podman VM. Probably something with code parts with `#ifdef __GNU__LIBRARY` or something like this. I will try to fix this.

**Review by hnarkaytis — APPROVED** [review]

**Pull request merged**